### PR TITLE
[NVPTX] Prefer prmt.b32 over bfi.b32

### PR DESCRIPTION
**Pull request by justinfargnoli**
In [[NVPTX] Improve lowering of v4i8](https://github.com/llvm/llvm-project/commit/cbafb6f2f5c99474164dcc725820cbbeb2e02e14) @Artem-B add the ability to lower ISD::BUILD_VECTOR with bfi PTX instructions. @Artem-B did this because: ([source](https://github.com/llvm/llvm-project/pull/67866#discussion_r1343066911))

> Under the hood byte extraction/insertion ends up as BFI/BFE instructions, so we may as well do that in PTX, too. https://godbolt.org/z/Tb3zWbj9b

However, the example that @Artem-B linked was targeting sm_52. On modern architectures, ptxas uses prmt.b32. [Example](https://godbolt.org/z/Ye4W1n84o).

Thus, remove uses of NVPTXISD::BFI in favor of NVPTXISD::PRMT.

### CODE REVIEW

**Comment by Artem-B** [comment]
Indeed. According to https://arxiv.org/pdf/2208.11174 `BFI` is much more expensive than `PRMT` which appears to take just 1 cycle on A100:
image

**Comment by Artem-B** [comment]
@justinfargnoli -- any chance NVIDIA could publish more details on the instruction timings. It's kind of hard to tell how fast something is expected to be by looking at the mnemonics alone.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 1]
```diff
@@ -2328,20 +2328,23 @@ SDValue NVPTXTargetLowering::LowerBUILD_VECTOR(SDValue Op,
...
+          {DAG.getAnyExtOrTrunc(Op->getOperand(0), DL, MVT::i32),
+           DAG.getAnyExtOrTrunc(Op->getOperand(1), DL, MVT::i32),
+           DAG.getConstant(0x3340, DL, MVT::i32),
+           DAG.getConstant(NVPTX::PTXPrmtMode::NONE, DL, MVT::i32)});
+      SDValue PRMT_210 = DAG.getNode(
+          NVPTXISD::PRMT, DL, MVT::v4i8,
+          {PRMT__10, DAG.getAnyExtOrTrunc(Op->getOperand(2), DL, MVT::i32),
+           DAG.getConstant(0x3410, DL, MVT::i32),
+           DAG.getConstant(NVPTX::PTXPrmtMode::NONE, DL, MVT::i32)});
+      SDValue PRMT3210 = DAG.getNode(
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2342]: Converting from `v = prmt(d, prmt(c, prmt(a,b)))` to `v = prmt(prmt(c,d), prmt(a,b))` may squeeze a bit more performance here if GPU can do two leaf permutes in parallel as they are independent.

Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2343]: Good point! 

I've implemented this in [5b1853f](https://github.com/llvm/llvm-project/pull/110766/commits/5b1853ff326c5701fd47e281bfc076e85bedb644).

**Comment by justinfargnoli** [comment]
> According to https://arxiv.org/pdf/2208.11174 BFI is much more expensive than PRMT which appears to take just 1 cycle on A100:

I think this refers to `bfi` in the general case. When the `c` and `d` operands are multiples of 8, `bfi` can be run as a `prmt`.

**Review by Artem-B — APPROVED** [review]
Please wait till Monday to land it, just in case.

**Comment by Artem-B** [comment]
> > According to https://arxiv.org/pdf/2208.11174 BFI is much more expensive than PRMT which appears to take just 1 cycle on A100:
> 
> I think this refers to `bfi` in the general case. When the `c` and `d` operands are multiples of 8, `bfi` can be run as a `prmt`.

Ptxas actully does compile such `bfi` into `prmt`: https://godbolt.org/z/9xTrhG99v
This implies that this patch is likely a no-op for that actual GPU code on the SASS level. 

Do we still want or need it?

**Comment by kalxr** [comment]
> Converting from v = prmt(d, prmt(c, prmt(a,b))) to v = prmt(prmt(c,d), prmt(a,b)) may squeeze a bit more performance here if GPU can do two leaf permutes in parallel as they are independent. 

I agree that this would be better for performance, and it doesn't like like ptxas does this.

> Do we still want or need it?

Given that this change would include the above improvement, IMO it would be useful.

**Comment by justinfargnoli** [comment]
> This implies that this patch is likely a no-op for that actual GPU code on the SASS level.
> Do we still want or need it?

CC @AlexMaclean, as he's voiced similar concerns offline. 

My original intention in creating this PR was to implement what I thought you had wanted to do in [[NVPTX] Improve lowering of v4i8](https://github.com/llvm/llvm-project/commit/cbafb6f2f5c99474164dcc725820cbbeb2e02e14) based on [your comment](https://github.com/llvm/llvm-project/pull/67866#discussion_r1343066911). 

But, as @kalxr has pointed out, this PR includes the `prmt(d, prmt(c, prmt(a,b))) --> prmt(prmt(c,d), prmt(a,b))` change that makes this more performant than the current approach. 

I'm also planning on doing more optimization work for `prmt`. This change would allow this lowering to take advantage of those.

**Comment by Artem-B** [comment]
I actually like ptmt better than bfi, as it also allows us to handle sign extension. We could use it as a base or example for cleaning up other i8/i16/i32 conversions later.

I think you're good to go.

**Review by kalxr — APPROVED** [review]

**Comment by Artem-B** [comment]
A logical follow-up to this patch would be to replace bfe -> prmt, too for byte extraction.

We should also take a look at v4i8 <-> v4i16 conversions. If we're not using prmt there, we should.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 2]
```diff
@@ -2328,20 +2328,23 @@ SDValue NVPTXTargetLowering::LowerBUILD_VECTOR(SDValue Op,
...
+          NVPTXISD::PRMT, DL, MVT::v4i8,
+          {DAG.getAnyExtOrTrunc(Op->getOperand(0), DL, MVT::i32),
+           DAG.getAnyExtOrTrunc(Op->getOperand(1), DL, MVT::i32),
+           DAG.getConstant(0x3340, DL, MVT::i32),
+           DAG.getConstant(NVPTX::PTXPrmtMode::NONE, DL, MVT::i32)});
+      SDValue PRMT32__ = DAG.getNode(
+          NVPTXISD::PRMT, DL, MVT::v4i8,
+          {DAG.getAnyExtOrTrunc(Op->getOperand(2), DL, MVT::i32),
+           DAG.getAnyExtOrTrunc(Op->getOperand(3), DL, MVT::i32),
+           DAG.getConstant(0x4033, DL, MVT::i32),
```
Comment (akuegel) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2341]: Looks like the mode here needs to be 0x3340, then our tests pass.

**Comment by cota** [comment]
Thanks @akuegel for pointing that out. I will revert this now since that confirms that this commit is the root cause of the failures we are seeing in our downstream tests (gross numerical disparities), despite no upstream tests being broken.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 2 above)
Comment (akuegel) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2341]: @justinfargnoli for visibility

**Pull request merged**